### PR TITLE
Added net8.0 and net9.0 target + tests

### LINKED ISSUE #3: Upgrade test projects to use net48, net8.0, and net9.0
This project's tests are targeting many frameworks that are out of support. We should upgrade the SDK to .NET 9 and test on `net48`, `net8.0`, and `net9.0`.

### PR DESCRIPTION
**Pull request by NightOwl888**
> NOTE: Please squash this PR

Closes #3.

- Added target and tests for `net8.0` and `net9.0`
- Removed `netcoreapp3.1` and `net5.0` tests
- Changed .NET Framework target from `net451` to `net462`
- Changed from using `net461` to `net48` for testing .NET Framework
- Added tests for `netstandard2.0` using `net471`
- Fixed issues with encoding provider loading on .NET Framework (see https://github.com/apache/lucenenet/issues/1025)
- Bumped dependency versions to eliminate vulnerable transitive dependencies
- Fixed nullable build warnings
- Changed publish output to publish each project to an individual directory
- Upgraded HTTP requests from using `WebRequest` to `HttpClient` and added tests

### CODE REVIEW

File: src/Morfologik.Stemming/Dictionary.cs [hunk 1]
```diff
@@ -80,14 +87,18 @@ public static Dictionary Read(Uri dictURL)
                 throw new IOException("Couldn't construct relative feature map URL for: " + dictURL, e);
             }
 
-            var fsaRequest = (HttpWebRequest)WebRequest.Create(dictURL);
-            var expectedMetadataRequest = (HttpWebRequest)WebRequest.Create(expectedMetadataURL);
+            try
+            {
+                using var httpClient = new HttpClient();
+                using var fsaStream = httpClient.GetStreamAsync(dictURL).GetAwaiter().GetResult();
```
Comment (paulirwin) [inline src/Morfologik.Stemming/Dictionary.cs:93]: Would it be better to make this method async instead? While `GetStreamAsync` uses `.ConfigureAwait(false)` and thus should not risk deadlocking, it might be desirable for callers to be able to await this method. We should make sure to use `.ConfigureAwait(false)` on the calls if this is made async.

File: src/Morfologik.Stemming/DictionaryMetadata.cs [hunk 2]
```diff
@@ -98,7 +99,7 @@ public sealed class DictionaryMetadata
         /// <summary>
         /// Gets all metadata attributes.
         /// </summary>
-        public IDictionary<DictionaryAttribute, string> Attributes => attributes.AsReadOnly();
+        public IDictionary<DictionaryAttribute, string> Attributes => JCG.Extensions.DictionaryExtensions.AsReadOnly(attributes);
```
Comment (paulirwin) [inline src/Morfologik.Stemming/DictionaryMetadata.cs:102]: Should this be of type `IReadOnlyDictionary<DictionaryAttribute, string>`? Would be a breaking change, so could be a separate PR/issue.

File: src/Morfologik.Stemming/Dictionary.cs (hunk 1 above)
Comment (paulirwin) [reply src/Morfologik.Stemming/Dictionary.cs:93]: Would be a breaking change to do so, so could/should be a separate PR/issue.

File: src/Morfologik.TestFramework/TypeExtensions.cs [hunk 3]
```diff
@@ -11,7 +11,7 @@ public static class TypeExtensions
         /// </summary>
         public static Stream getResourceAsStream(this Type t, string name)
         {
-            return t.Assembly.FindAndGetManifestResourceStream(t, name);
+            return t.Assembly.FindAndGetManifestResourceStream(t, name)!;
```
Comment (paulirwin) [inline src/Morfologik.TestFramework/TypeExtensions.cs:14]: Perhaps it would be better here to coalesce this with a more meaningful exception than returning null, i.e. `?? throw new InvalidOperationException($"Could not find manifest resource stream '${name}'.");

File: .build/runbuild.ps1 [hunk 4]
```diff
@@ -140,6 +140,11 @@ task Test -depends Pack -description "This task runs the tests" {
             $testPlatformArray = $testPlatforms -split '\s*[;,]\s*'
             foreach ($testPlatform in $testPlatformArray) {
 
+                if ([System.Runtime.InteropServices.RuntimeInformation]::OSDescription -match "Darwin" -and $testPlatform -eq "arm64" -and $framework -eq "net5.0") {
```
[the remark this answers is not in the document]
Comment (NightOwl888) [reply .build/runbuild.ps1:143]: The `net5.0` tests have been removed. However, we use the same runbuild.ps1 (except for a few lines) in several projects, and those other projects may still target `net5.0`.

File: src/Morfologik.Stemming/Dictionary.cs (hunk 1 above)
Comment (NightOwl888) [reply src/Morfologik.Stemming/Dictionary.cs:93]: Good idea. Looks like the protocol for maintaining existing synchronous methods for backward compatibility is to add a new method named `ReadAsync()`

```c#
/// <summary>
/// Attempts to load a dictionary using the URL to the FSA file and the
/// expected metadata extension asynchronously.
/// </summary>
/// <param name="dictURL">The URL pointing to the dictionary file (<c>*.dict</c>).</param>
/// <returns>A task representing the asynchronous operation. The result is an instantiated dictionary.</returns>
/// <exception cref="IOException">If an I/O error occurs.</exception>
public static async Task<Dictionary> ReadAsync(Uri dictURL)
{
    Uri expectedMetadataURL;
    try
    {
        string external = dictURL.AbsoluteUri;
        expectedMetadataURL = new Uri(DictionaryMetadata.GetExpectedMetadataFileName(external));
    }
    catch (UriFormatException e)
    {
        throw new IOException("Couldn't construct relative feature map URL for: " + dictURL, e);
    }

    try
    {
        using var httpClient = new HttpClient();
        using var fsaStream = await httpClient.GetStreamAsync(dictURL).ConfigureAwait(false);
        using var metadataStream = await httpClient.GetStreamAsync(expectedMetadataURL).ConfigureAwait(false);

        return Read(fsaStream, metadataStream);
    }
    catch (HttpRequestException e)
    {
        throw new IOException("Error retrieving data from the specified URLs.", e);
    }
}
```

And then call it from the synchronous method

```c#
public static Dictionary Read(Uri dictURL)
{
    return ReadAsync(dictURL).GetAwaiter().GetResult();
}
```

Thoughts?

File: src/Morfologik.Stemming/DictionaryMetadata.cs (hunk 2 above)
Comment (NightOwl888) [reply src/Morfologik.Stemming/DictionaryMetadata.cs:102]: Unfortunately, while all BCL types that implement `IDictionary<TKey, TValue>` also implement `IReadOnlyDictionary<TKey, TValue>` there is no guarantee this will always be the case. So, casting is not guaranteed to work, and we need `IDictionary<TKey, TValue>` to call other APIs. Basically, this is an all or nothing proposition. We would need to change all APIs to be `IReadOnlyDictionary<TKey, TValue>`.

Also, this component is already released and there has only been 1 patch released higher than 2.1.7. So, we would need to introduce breaking changes in a patch. Not saying that we shouldn't, but it wouldn't be very nice if we had breaking changes (believe it or not, there are other libraries depending on this other than Lucene.NET, such as https://www.nuget.org/packages/Clara.Analysis.Morfologik#dependencies-body-tab).

> NOTE: I just noticed that there are 3 modules of this that haven't been ported [`morfologik-fsa-builders`](https://github.com/morfologik/morfologik-stemming/tree/master/morfologik-fsa-builders), [`morfologik-speller`](https://github.com/morfologik/morfologik-stemming/tree/master/morfologik-speller), and [`morfologik-tools`](https://github.com/morfologik/morfologik-stemming/tree/master/morfologik-tools). I don't recall whether there was a reason for that other than the fact that Lucene.NET doesn't directly depend on them.

File: src/Morfologik.TestFramework/TypeExtensions.cs (hunk 3 above)
Comment (NightOwl888) [reply src/Morfologik.TestFramework/TypeExtensions.cs]: Good idea. I added an exception.

File: src/Morfologik.Stemming/DictionaryMetadata.cs (hunk 2 above)
Comment (paulirwin) [reply src/Morfologik.Stemming/DictionaryMetadata.cs:102]: Good point re: breaking change, I think that could be split out as an issue for future improvement.

File: src/Morfologik.Stemming/Dictionary.cs (hunk 1 above)
Comment (paulirwin) [reply src/Morfologik.Stemming/Dictionary.cs:93]: Oh, IAsyncDisposable might not be supported by all targets. If not, then nevermind on that one.

File: src/Morfologik.Stemming/DictionaryMetadata.cs (hunk 2 above)
Comment (NightOwl888) [reply src/Morfologik.Stemming/DictionaryMetadata.cs:102]: Well, in this case we could do it because there are specific closing types so we can see where in Lucene.NET this is valid to do. But in general, we have been avoiding read only collection interfaces for compatibility reasons. We cannot always make assumptions about where the Lucene designers intended the output of one member to be used in the input of another.

Comment (NightOwl888) [reply src/Morfologik.Stemming/DictionaryMetadata.cs:102]: I opened #6, #7, #8, and #9 to deal with these.

Comment (NightOwl888) [reply src/Morfologik.Stemming/DictionaryMetadata.cs:102]: I opened #7, #8, and #9 to deal with these.

File: src/Morfologik.Stemming/Dictionary.cs (hunk 1 above)
Comment (NightOwl888) [reply src/Morfologik.Stemming/Dictionary.cs:93]: I opened #6 to deal with this.
